File: java/AASK/src/root/MatchParser.java
package root;

/**
 *
 * @author Curtis Fenner
 */
import java.io.*;
import java.net.URL;
import java.util.Hashtable;

import Jama.Matrix;

public class MatchParser {
	//// CONSTANTS ------------------------------------------------------------

	public static final String MATCH_RESULTS = "matchresults.html";
	public static final String TEAM_STANDINGS = "rankings.html";
	
	//// PRIVATE VARIABLES ----------------------------------------------------
	
	private int _teamCount;
	private int _matchesPlayed;
	private int _matchesTotal;
	
	private String _competitionURL;
	
	// Match Results
	private Table _matchResults;
	private Matrix _matchResultsMatrix;
	
	// Team Standings
	private Table _teamStandings;
	private Matrix _teamStandingsMatrix;
	
	// Hashtable
	private Hashtable<Integer, Integer> _teamHash;
	
	// Auxiliary Matrices
	private Matrix _matchPairFrequencies;
	private Matrix _autonPoints;
	private Matrix _climbPoints;
	private Matrix _teleopPoints;
	private Matrix _totalPoints;
	
	// Estimated Contributions
	private Matrix _autonContributions;
	private Matrix _climbContributions;
	private Matrix _teleopContributions;
	private Matrix _totalContributions;
	
	//// CONSTRUCTOR ----------------------------------------------------------
	
	public MatchParser(){
		
	}
	
	public MatchParser(String competitionURL){
		connect(competitionURL);
	}
	
	//// CONNECT --------------------------------------------------------------
	
	public void connect(String competitionURL){
		_competitionURL = competitionURL;
		parse();
	}
	
	//// PARSING --------------------------------------------------------------
	
	public void parse(){
		// Verify Arguments
		if(_competitionURL == null) throw new NullPointerException("MatchParser URL cannot be null.");
		
		// Parse
		_matchResults = new Table(_competitionURL + MATCH_RESULTS);
		_teamStandings = new Table(_competitionURL + TEAM_STANDINGS);
		
		// Construct Hashtable (Team # -> Rank)
		_teamHash = _teamStandings.constructColumnHash(1, 0);
		
		// Dataset Properties
		_teamCount = _teamHash.size();
		_matchesTotal = _matchResults.getRows();
		
		// Calculate Match Pair Frequencies
		double[][] mpf = new double[_teamCount][_teamCount];
		double[][] matchResultsBody = _matchResults.getBody();
		
		Matrix mrb = new Matrix(matchResultsBody);
		mrb.print(2, 0);
		
		for(int i = 0; i < _matchesTotal; i++){ 	// Matches
			if(matchResultsBody[8][i] == 0 && matchResultsBody[9][i] == 0) continue;
			for(int j = 2; j <= 4; j++){			// Select Row Team
				for(int k = j; k <= 4; k++){		// Select Column Team
					// Red Alliance
					int rowTeam = (int)matchResultsBody[j][i];
					int rowTeamIndex = _teamHash.get(rowTeam) - 1;
					int colTeam = (int)matchResultsBody[k][i];
					int colTeamIndex = _teamHash.get(colTeam) - 1;
					
					mpf[rowTeamIndex][colTeamIndex]++;
					mpf[colTeamIndex][rowTeamIndex]++;
				
					// Blue Alliance
					rowTeam = (int)matchResultsBody[j+3][i];
					colTeam = (int)matchResultsBody[k+3][i];
					rowTeamIndex = _teamHash.get(rowTeam) - 1;
					colTeamIndex = _teamHash.get(colTeam) - 1;
					
					mpf[rowTeamIndex][colTeamIndex]++;
					mpf[colTeamIndex][rowTeamIndex]++;
				}
			}
		}
		
		for(int i = 0; i < _teamCount; i++){
			int sum = 0;
			for(int j = 0; j < _teamCount; j++){
				if(i == j) continue;
				sum += mpf[j][i];
			}
			mpf[i][i] = sum / 2;
		}
		
		_matchPairFrequencies = new Matrix(mpf);
		
		// Extract Point Values
		_teamStandingsMatrix = new Matrix(_teamStandings.getBody()).transpose();
		_teamStandingsMatrix.print(3, 0);
		_autonPoints = _teamStandingsMatrix.getMatrix(0,39,3,3);
		_climbPoints = _teamStandingsMatrix.getMatrix(0,39,4,4);
		_teleopPoints = _teamStandingsMatrix.getMatrix(0,39,5,5);
		_totalPoints = _autonPoints.plus(_climbPoints.plus(_teleopPoints));
	}
	
	//// GETTER / SETTER METHODS ----------------------------------------------
        
        public Matrix getTeamStandingsMatrix() {
            return this._teamStandingsMatrix;
        }
	
	public Hashtable getTeamHash(){
		return _teamHash;
	}
	
	public int getTeamCount(){
		return _teamCount;
	}
	
	public Table getMatchResults(){
		return _matchResults;
	}
	
	public Table getTeamStandings(){
		return _teamStandings;
	}
	
	public Matrix getAutonPoints(){
		return _autonPoints;
	}
	
	public Matrix getClimbPoints(){
		return _climbPoints;
	}
	
	public Matrix getTeleopPoints(){
		return _teleopPoints;
	}
	
	public Matrix getTotalPoints(){
		return _totalPoints;
	}
	
	public Matrix getAutonContributions(){
		if(_autonContributions != null) return _autonContributions;
		else return (_autonContributions = _matchPairFrequencies.solve(_autonPoints));
	}
	
	public Matrix getClimbContributions(){
		if(_climbContributions != null) return _climbContributions;
		else return (_climbContributions = _matchPairFrequencies.solve(_climbPoints));
	}
	
	public Matrix getTeleopContributions(){
		if(_teleopContributions != null) return _teleopContributions;
		else return (_teleopContributions = _matchPairFrequencies.solve(_teleopPoints));
	}
	
	public Matrix getTotalContributions(){
		if(_totalContributions != null) return _totalContributions;
		else return (_totalContributions = _matchPairFrequencies.solve(_totalPoints));
	}
	
}

File: web/matchParser.js
/**
 * MATCHPARSER.JS
 * Designed as the equivalent of MatchParser.java, which essentially holds data
 * and calculations for a single FIRST event.  It is responsible for requesting
 * and processing match data.
 * @author Ben Bray
 */

//// MATCH PARSER CONSTRUCTOR -------------------------------------------------

function MatchParser(rankingsURL, resultsURL){
	var mp = {type:"MatchParser"};
	
	// Fields
	mp.rankingsURL = rankingsURL;
	mp.resultsURL = resultsURL;
	
	// Methods
	mp.parse = function(onparsingcomplete) {
		mp.onparsingcomplete = onparsingcomplete;
		MatchParserBeginParse(mp);
	};
	
	// Return Match Parser
	return mp;
}

function MatchParserTypeCheck(mp){
	// Type Check
	if(!mp || !mp.type || !mp.type == "MatchParser"){
		throw "MatchParser :: Illegal Argument : mp is not of type 'MatchParser'";
	}
}

//// MATCH PARSER METHODS -----------------------------------------------------

function MatchParserBeginParse(mp){
	console.log("\t[MatchParserBeginParse] started");
	// Type Check
	MatchParserTypeCheck(mp);
	
	// Verify URL Existence
	if(!mp.rankingsURL || !mp.resultsURL){
		throw "MatchParserParse :: Undefined URL";
	}
	
	// Create Tables
	mp.resultsTable = Table(mp.resultsURL);
	mp.rankingsTable = Table(mp.rankingsURL);
	
	// Begin Loading Results
	MatchParserLoadResults(mp);
}

/**
 * Loads the Results Table, and afterwards either loads the rankings table or
 * proceeds to the MatchParserCompleteParse() method to perform calculations.
 * @mp The MatchParser object to use.
 */
function MatchParserLoadResults(mp){
	console.log("[MatchParserLoadResults] started");
	// Type Check
	MatchParserTypeCheck(mp);
	
	if(!mp.resultsTable.requestComplete){
		if(!mp.rankingsTable.requestComplete){
			mp.resultsTable.loadTable(function() { MatchParserLoadRankings(mp); });
		} else {
			mp.resultsTable.loadTable(MatchParserCompleteParse);
		}
	}
	console.log("[MatchParserLoadResults] success");
}

/**
 * Loads the Rankings Table, and afterwards either loads the results table or
 * proceeds to the MatchParserCompleteParse() method to perform calculations.
 * @mp The MatchParser object to use.
 */
function MatchParserLoadRankings(mp){
	console.log("[MatchParserLoadRankings] started");
	// Type Check
	MatchParserTypeCheck(mp);
	
	if(!mp.rankingsTable.requestComplete){
		if(!mp.resultsTable.requestComplete){
			mp.rankingsTable.loadTable(function() { MatchParserLoadResults(mp); });
		} else {
			mp.rankingsTable.loadTable(function() { MatchParserCompleteParse(mp); });
		}
	}
	console.log("[MatchParserLoadRankings] success");
}

/**
 * Completes the parsing process once all Tables have been successfully loaded.
 * @mp The MatchParser object to use.
 */
function MatchParserCompleteParse(mp){
	console.log("[MatchParserCompleteParse] started");
	// Type Check
	MatchParserTypeCheck(mp);
	
	// Get Table Data
	resultsBody = mp.resultsTable.getBody();
	rankingsBody = mp.rankingsTable.getBody();
	
	teamHash = mp.rankingsTable.constructColumnHash(1, 0);
	teamCount = teamHash.size();
	matchesTotal = mp.resultsTable.getRows();
	
	// Count Match Pair Frequencies
	mpf = zeros(teamCount, teamCount);
	
	for(var i = 0; i < matchesTotal; i++){
		if(resultsBody[8][i] == 0 && resultsBody[9][i] == 0) continue;
		for(var j = 2; j <= 4; j++){			// Select Row Team
			for(var k = j; k <= 4; k++){		// Select Column Team
				// Red Alliance
				var rowTeam = resultsBody[j][i];
				var rowTeamIndex = teamHash.get(rowTeam) - 1;
				var colTeam = resultsBody[k][i];
				var colTeamIndex = teamHash.get(colTeam) - 1;
				
				mpf[rowTeamIndex][colTeamIndex]++;
				if(colTeamIndex != rowTeamIndex) { mpf[colTeamIndex][rowTeamIndex]++; };
			
				// Blue Alliance
				rowTeam = resultsBody[j+3][i];
				colTeam = resultsBody[k+3][i];
				rowTeamIndex = teamHash.get(rowTeam) - 1;
				colTeamIndex = teamHash.get(colTeam) - 1;
				
				mpf[rowTeamIndex][colTeamIndex]++;
				if(colTeamIndex != rowTeamIndex) { mpf[colTeamIndex][rowTeamIndex]++; };
			}
		}
	}
	
	// Correct Diagonal Sum
	for(var i = 0; i < teamCount; i++){
		var sum = 0;
		for(var j = 0; j < teamCount; j++){
			if(i == j) continue;
			sum += mpf[j][i];
		}
		mpf[i][i] = sum / 2;
	}
	
	// Set Object Fields
	mp.resultsBody = resultsBody;
	mp.rankingsBody = rankingsBody;
	mp.teamHash = teamHash;
	mp.teamCount = teamCount;
	mp.matchesTotal = matchesTotal;
	mp.mpf = mpf;
	
	// Extract Point Values
	mp.autonPoints = new Array(teamCount);
	mp.climbPoints = new Array(teamCount);
	mp.teleopPoints = new Array(teamCount);
	mp.totalPoints = new Array(teamCount);
	
	for(var i = 0; i < teamCount; i++){
		mp.autonPoints[i] = parseInt(rankingsBody[3][i]);
		mp.climbPoints[i] = parseInt(rankingsBody[4][i]);
		mp.teleopPoints[i] = parseInt(rankingsBody[5][i]);
		mp.totalPoints[i] = mp.autonPoints[i] + mp.climbPoints[i] + mp.teleopPoints[i];
	}
	
	// Call 'onparsingcomplete'
	mp.onparsingcomplete.call();
}
